vault/logical_system.go

```diff
@@ -2752,19 +2751,39 @@ func (b *SystemBackend) pathRandomWrite(ctx context.Context, req *logical.Reques
 	return resp, nil
 }
 
+// hasMountAccess returns true if path has a non-deny capability in acl,
+// or if there exist any non-deny capability beneath that path in acl.
 func hasMountAccess(ctx context.Context, acl *ACL, path string) bool {
+	return hasNonDenyCapability(ctx, acl, path) || containsNonDenyCapability(ctx, acl, path)
+}
+
+// hasNonDenyCapability returns true if path has a non-deny capability in acl.
+func hasNonDenyCapability(ctx context.Context, acl *ACL, path string) bool {
 	ns, err := namespace.FromContext(ctx)
 	if err != nil {
 		return false
 	}
 
-	// If an earlier policy is giving us access to the mount path then we can do
-	// a fast return.
-	capabilities := acl.Capabilities(ctx, ns.TrimmedPath(path))
-	if !strutil.StrListContains(capabilities, DenyCapability) {
+	req := &logical.Request{
+		Path: ns.TrimmedPath(path),
+		// doesn't matter, but use List to trigger fallback behavior so we can
+		// model real behavior
+		Operation: logical.ListOperation,
+	}
+
+	res := acl.AllowOperation(ctx, req, true)
+	if res.IsRoot {
 		return true
 	}
+	if res.CapabilitiesBitmap&DenyCapabilityInt > 0 {
+		return false
+	}
+	return res.CapabilitiesBitmap > 0
+}
 
+// containsNonDenyCapability returns true if there exist any non-deny capability
+// beneath that path in acl.
+func containsNonDenyCapability(ctx context.Context, acl *ACL, path string) bool {
```

Comment:
I wonder if we can come up with more clear names for these two functions? Maybe `hasNonDenyCapability` and `hasAccessToSubtreePaths`? or maybe `hasAccess` and `hasAccessChildren`?

Reply:
Both of those sound fine, I'll try them out and see which I like better.